pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -19,28 +19,526 @@ package service
+				if ok {
+					r.pods.Store(pods)
+				} else {
+					log.Infof("pod not found: %s", r.agentName)
+				}
+			}),
+		)),
+		k8s.WithResourceController(mpod.New(
+			mpod.WithControllerName("pod metrics discover"),
+			mpod.WithOnErrorFunc(func(err error) {
+				log.Error(err)
+			}),
+			mpod.WithOnReconcileFunc(func(podList map[string]mpod.Pod) {
+				if len(podList) > 0 {
+					r.podMetrics.Store(podList)
+				} else {
+					log.Info("pod metrics not found")
+				}
+			}),
+		)),
+		k8s.WithResourceController(cm),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	r.decoder, err = decoder.NewDecoder(r.ctrl.GetManager().GetScheme())
+	if err != nil {
+		return nil, err
+	}
+
 	return r, nil
 }
 
+// Start starts the rebalancer controller loop for the Vald agent index rebalancer.
 func (r *rebalancer) Start(ctx context.Context) (<-chan error, error) {
-	ech := make(chan error, 2)
+	cech, err := r.ctrl.Start(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	ech := make(chan error, 1)
+	r.eg.Go(safety.RecoverFunc(func() error {
+		defer close(ech)
+		dt := time.NewTicker(r.rcd)
+		defer dt.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-dt.C:
+				var (
+					ss statefulset.StatefulSet
+					ok bool
+
+					podModels       map[string][]*model.Pod
+					namespaceByJobs map[string][]job.Job
+					ssModel         map[string]*model.StatefulSet
+					jobTpl          *job.Job
+				)
+
+				podModels, err := r.genPodModels()
+				if err != nil {
+					log.Infof("error generating pod models: %s", err.Error())
+					continue
+				}
+
+				namespaceByJobs, err = r.namespaceByJobs()
+				if err != nil {
+					log.Infof("error generating job models: %s", err.Error())
+				}
+
+				jobTpl, err = r.genJobTpl()
+				if err != nil {
+					log.Infof("error generating job template: %s", err.Error())
+					continue
+				}
+
+				// TODO: cache specified reconciled result based on agentResourceType.
+				switch r.agentResourceType {
+				case config.STATEFULSET:
+					ss, ok = r.statefulSets.Load().(statefulset.StatefulSet)
+					if !ok {
+						log.Info("statefulset is empty")
+						continue
+					}
+					ssModel = make(map[string]*model.StatefulSet)
+					ssModel[ss.Namespace] = &model.StatefulSet{
+						Name:            ss.Name,
+						Namespace:       ss.Namespace,
+						DesiredReplicas: ss.Spec.Replicas,
+						Replicas:        ss.Status.Replicas,
+					}
+
+					for ns, sm := range ssModel {
+						if *sm.DesiredReplicas != sm.Replicas {
+							log.Debugf("[not desired] desired replica: %d, current replica: %d", *sm.DesiredReplicas, sm.Replicas)
+							continue
+						}
+
+						maxPodName, rate := r.getBiasOverDetail(podModels[ns])
+						if maxPodName == "" || rate < r.rateThreshold {
+							log.Debugf("[rate/podname checking] pod name, rate, rateThreshold: %s, %.3f, %f", maxPodName, rate, r.rateThreshold)
+							continue
+						}
+						// log.Debugf("[bias/jobcheck] jobName: %#v", [r.jobNamespace])
```

Comment:
debug log???

Reply:
maybe we can delete this line?
@hlts2 @vankichi

Reply:
```suggestion
```

Reply:
thank you.
I fixed: https://github.com/vdaas/vald/pull/1187/commits/dc9370a0336acf1c19289d46a4d9b6042beec548